_posts/2025-12-26-Deploying_OCI_Images_with_Custom_IPA_Hardware_Manager.md

```diff
@@ -0,0 +1,1966 @@
+        - gateway:
+            fromIPPool: my-ip-pool
+          network: 0.0.0.0
+          prefix: 0
+    services:
+      dns:
+      - 8.8.8.8
+```
+
+## Building an IPA Image with the Hardware Manager
+
+To use this hardware manager, you need to build a custom IPA ramdisk
+image using
+[ironic-python-agent-builder](https://opendev.org/openstack/ironic-python-agent-builder).
+This tool uses [diskimage-builder](https://docs.openstack.org/diskimage-builder/latest/)
+(DIB) to create bootable ramdisk images containing the IPA and any
+custom elements you need.
+
+### Required Packages
+
+The hardware manager requires several packages to be present in the
+IPA ramdisk:
+
+| Package | Purpose |
+|---------|---------|
+| `crane` | OCI image extraction from container registries |
+| `mdadm` | Software RAID array management |
+| `lvm2` | Logical Volume Manager for root filesystem |
+| `parted` | Disk partitioning |
+| `dosfstools` | FAT32 filesystem creation for EFI partition |
+| `grub2-efi-*` | UEFI bootloader installation |
+| `curl` | Downloading files during deployment |
+| `rsync` | EFI partition synchronization for RAID |
+
+### Custom DIB Elements
+
+DIB elements are modular components that customize the image build.
+Each element is a directory containing scripts that run at different
+phases of the build:
+
+| Directory | Phase | Description |
+|-----------|-------|-------------|
+| `extra-data.d/` | Pre-build | Copy files into build environment |
+| `install.d/` | Chroot | Run inside chroot during build |
+| `post-install.d/` | Post-install | Run after package installation |
+| `finalise.d/` | Finalize | Run at end of build process |
+
+Scripts are named with a numeric prefix (e.g., `50-crane`) to control
+execution order.
+
+<!-- markdownlint-disable MD033 -->
+
+<details>
+  <summary>DIB element: crane (OCI image tool)</summary>
+  <div markdown="1">
+
+<!-- markdownlint-enable MD033 -->
+
+Create a DIB element to install Google's `crane` tool for OCI image
+extraction. Create the following directory structure:
+
+```text
+crane/
+├── element-deps
+└── install.d/
+    └── 50-crane
+```
+
+The `element-deps` file can be empty or list dependencies. The install
+script (`install.d/50-crane`):
+
+```bash
+#!/bin/bash
+
+# https://docs.openstack.org/diskimage-builder/latest/developer/developing_elements.html
+
+if [ "${DIB_DEBUG_TRACE:-0}" -gt 0 ]; then
+    set -x
+fi
+
+set -eu
+set -o pipefail
+
+CRANE_VERSION="${DIB_CRANE_VERSION:-latest}"
+
+# Detect architecture
+ARCH=$(uname -m)
+case "${ARCH}" in
+    x86_64)
+        CRANE_ARCH="x86_64"
+        ;;
+    aarch64)
+        CRANE_ARCH="arm64"
+        ;;
+    *)
+        echo "Unsupported architecture: ${ARCH}"
+        exit 1
+        ;;
+esac
+
+echo "Installing crane (${CRANE_VERSION}) for ${CRANE_ARCH}..."
+
+# Get the download URL
+if [ "${CRANE_VERSION}" = "latest" ]; then
+    DOWNLOAD_URL=$(curl -s https://api.github.com/repos/google/go-containerregistry/releases/latest |
```

Comment:
I'm curious if you considered using podman, which is readily available in CS9 and also has the export command.

Reply:
hmm, I guess that can also be an option